File: MATLAB - LIQUAC copy/MediumRange.m
classdef MediumRange % based on LIQUAC model
    
    methods

        % Medium range activity coefficients, no temperature dependence, as
        % specified by Kiepe et al. LIQUAC
        function ln_gamma_MR = func_MR(obj, b_ij, c_ij, b_jcja, c_jcja, x, MW, valency, salt)
            
            % Initializes key variables and interaction terms B_ij
            [m_ion, ionic_strength, x_saltfree, mean_MW] = variables(obj, x, MW, valency);
            [B_solv, dB_solv, B_ion, dB_ion] = interactionTerms(obj, b_ij, c_ij, b_jcja, c_jcja, ionic_strength, salt);
        
            % Calculates solvent and ionic activity coefficients
            [ln_gamma_solv_MR] = solventGamma(obj, m_ion, x_saltfree, ionic_strength, MW, mean_MW, B_solv, dB_solv, B_ion, dB_ion);
            [ln_gamma_ions_MR] = ionsGamma(obj, mean_MW, m_ion, x_saltfree, B_solv, dB_solv, B_ion, dB_ion, MW, b_ij, c_ij, valency);
            
            ln_gamma_MR = [ln_gamma_solv_MR, ln_gamma_ions_MR];            
        end

        % Medium range activity coefficients, as specified by Kiepe et al.
        % LIQUAC, with temperature dependence added into the interaction
        % terms
        function ln_gamma_MR = func_MR_T(obj, b_ij, c_ij, b_jcja, c_jcja, x, MW, valency, salt, T)
            
            % Initializes key variables and interaction terms B_ij
            [m_ion, ionic_strength, x_saltfree, mean_MW] = variables(obj, x, MW, valency);
            [B_solv, dB_solv, B_ion, dB_ion] = interactionTerms_TempDependency(obj, b_ij, c_ij, b_jcja, c_jcja, ionic_strength, T, salt);
        
            % Calculates solvent and ionic activity coefficients
            [ln_gamma_solv_MR] = solventGamma(obj, m_ion, x_saltfree, ionic_strength, MW, mean_MW, B_solv, dB_solv, B_ion, dB_ion);
            [ln_gamma_ions_MR] = ionsGamma(obj, mean_MW, m_ion, x_saltfree, B_solv, dB_solv, B_ion, dB_ion, MW, b_ij, c_ij, valency);
            
            ln_gamma_MR = [ln_gamma_solv_MR, ln_gamma_ions_MR];                  
        end


        % Initializes key variables for medium range activity coefficients
        function [m_ion, ionic_strength, x_saltfree, mean_MW] = variables(obj, x, MW, valency)
            
            kg_solvent = x(:, 1) * MW(1) + x(:, 2) * MW(2);
            m_ion = x(:,3:4) ./ kg_solvent;
            ionic_strength = 0.5 * sum(m_ion .* valency(3:4).^2,2);
            x_saltfree = [x(:,1) ./ (x(:,1) + x(:,2)), x(:,2) ./ (x(:,1) + x(:,2))];
            mean_MW = x_saltfree(:,1) * MW(1) + x_saltfree(:,2) * MW(2);

        end

        % Calculates B_solv, dB_solv, B_ion, dB_ion, no temperature
        % dependence
        function [B_solv, dB_solv, B_ion, dB_ion] = interactionTerms(obj, b_ij, c_ij, b_jcja, c_jcja, ionic_strength, salt)
            % Set variables a1-a4 based on LIQUAC paper
            
            a_1 = -1.2;
            a_3 = -1;
            if salt == "LiCl"
                a_4 = 0.1451;
            elseif salt == "LiBr"
                a_4 = 0.0695;
            elseif salt == "LiI"
                a_4 = 0.1797;
            elseif salt == "LiOH"
                a_4 = 0.0894;
            elseif salt == "LiNO3"
                a_4 = 0.1820;
            elseif salt == "Li2SO4"
                a_4 = 0.2936;
            elseif salt == "LiClO3"
                a_4 = 0.1325;
            elseif salt == "CaCl2"
                a_4 = 0.2170;
            else
                a_4 = 0.1250;
            end
            a_2 = 2 * a_4;
            
            % [solvent-cation, solvent-anion, water-cation, water-anion]
            B_solv = b_ij + c_ij .* exp(a_1 * sqrt(ionic_strength) + a_2 * ionic_strength);
            dB_solv = (a_1 / 2 ./ sqrt(ionic_strength) + a_2) .* c_ij .* exp(a_1 * sqrt(ionic_strength) + a_2 * ionic_strength);
            B_ion = b_jcja + c_jcja .* exp(a_3 * sqrt(ionic_strength) + a_4 * ionic_strength);
            dB_ion = (a_3 / 2 ./ sqrt(ionic_strength) + a_4) .* c_jcja .* exp(a_3 * sqrt(ionic_strength) + a_4 * ionic_strength);            
        end

        % Calculates B_solv, dB_solv, B_ion, dB_ion, with temperature
        % dependence
        function [B_solv, dB_solv, B_ion, dB_ion] = interactionTerms_TempDependency(obj, b_ij, c_ij, b_jcja, c_jcja, ionic_strength, T, salt)
            
            % Set variables a1-a4 based on LIQUAC paper
            a_1 = -1.2;
            a_3 = -1;
            if salt == "LiCl"
                a_4 = 0.1451;
            elseif salt == "LiBr"
                a_4 = 0.0695;
            elseif salt == "LiI"
                a_4 = 0.1797;
            elseif salt == "LiOH"
                a_4 = 0.0894;
            elseif salt == "LiNO3"
                a_4 = 0.1820;
            elseif salt == "Li2SO4"
                a_4 = 0.2936;
            elseif salt == "LiClO3"
                a_4 = 0.1325;
            elseif salt == "CaCl2"
                a_4 = 0.2170;
            else
                a_4 = 0.1250;
            end
            a_2 = 2 * a_4;

            % [solv_1-cation, solv_1-anion, solv_2-cation, solv2_anion]
            B_solv = b_ij ./ T + c_ij ./ T .* exp(a_1 * sqrt(ionic_strength) + a_2 * ionic_strength);
            dB_solv = (a_1 / 2 ./ sqrt(ionic_strength) + a_2) .* c_ij ./ T .* exp(a_1 * sqrt(ionic_strength) + a_2 * ionic_strength);
            B_ion = b_jcja ./ T + c_jcja ./ T .* exp(a_3 * sqrt(ionic_strength) + a_4 * ionic_strength);
            dB_ion = (a_3 / 2 ./ sqrt(ionic_strength) + a_4) .* c_jcja ./ T .* exp(a_3 * sqrt(ionic_strength) + a_4 * ionic_strength);
           
        end

        % Calculates medium range activity coefficients for solvents
        function [ln_gamma_solv_MR] = solventGamma(obj, m_ion, x_saltfree, ionic_strength, MW, mean_MW, B_solv, dB_solv, B_ion, dB_ion)

            Solv1_Term1 = sum(m_ion .* B_solv(:,1:2),2);
            Solv1_Term2 = MW(1) ./ mean_MW .* (sum((x_saltfree(:,1) .* m_ion .* (B_solv(:,1:2) + ionic_strength .* dB_solv(:,1:2))),2) + sum((x_saltfree(:,2) .* m_ion .* (B_solv(:,3:4) + ionic_strength .* dB_solv(:,3:4))),2));
            
            Solv2_Term1 = sum(m_ion .* B_solv(:,3:4),2);
            Solv2_Term2 = MW(2) ./ mean_MW .* (sum((x_saltfree(:,1) .* m_ion .* (B_solv(:,1:2) + ionic_strength .* dB_solv(:,1:2))),2) + sum((x_saltfree(:,2) .* m_ion .* (B_solv(:,3:4) + ionic_strength .* dB_solv(:,3:4))),2));
            
            Solv_Term3 = (m_ion(:,1) .* m_ion(:,2)) .* (B_ion + ionic_strength .* dB_ion);
            
            ln_gamma_solv_MR = [Solv1_Term1 - Solv1_Term2 - MW(1) .* Solv_Term3, Solv2_Term1 - Solv2_Term2 - MW(2) .* Solv_Term3];           
        end

        % Calculates medium range activity coefficients for ions
        function [ln_gamma_ions_MR] = ionsGamma(obj, mean_MW, m_ion, x_saltfree, B_solv, dB_solv, B_ion, dB_ion, MW, b_ij, c_ij, valency)            
            
            Cation_Term1 = 1 ./ mean_MW .* sum(x_saltfree(:,1) .* B_solv(:,1) + x_saltfree(:,2) .* B_solv(:,3),2);
            Anion_Term1 = 1 ./ mean_MW .* sum(x_saltfree(:,1) .* B_solv(:,2) + x_saltfree(:,2) .* B_solv(:,4),2);
            
            Ions_Term2 = (valency(:,3:4).^2) ./ (2 .* mean_MW) .* (sum(x_saltfree(:,1) .* m_ion .* dB_solv(:,1:2),2) + sum(x_saltfree(:,2) .* m_ion .* dB_solv(:,3:4),2));
            Ions_Term3 = [m_ion(:,2) .* B_ion, m_ion(:,1) .* B_ion];
            Ions_Term4 = 0.5 * (valency(:,3:4).^2) .* sum(m_ion(:,1) .* m_ion(:,2).* dB_ion,2);

            % Reference states
            Ions_RefState_Cation = 1 ./ MW(2) .* (b_ij(:,3) + c_ij(:,3));
            Ions_RefState_Anion = 1 ./ MW(2) .* (b_ij(:,4) + c_ij(:,4));
            
            ln_gamma_ions_MR = [Cation_Term1 + Ions_Term2(:,1) + Ions_Term3(:,1) + Ions_Term4(:,1) - Ions_RefState_Cation, Anion_Term1 + Ions_Term2(:,2) + Ions_Term3(:,2) + Ions_Term4(:,2) - Ions_RefState_Anion];           
            
        end

        
    end
end

File: PYTHON - LIQUAC/MediumRange.py
"""
MediumRange.py
==============
Medium-range contribution to activity coefficients in the LIQUAC model
(Kiepe et al., 2006).

Species ordering: [solvent, water, cation, anion]
"""

import numpy as np


# Mapping from salt name to the a_4 constant used in the B_ion term.
_A4_MAP = {
    "LiCl":   0.1451,
    "LiBr":   0.0695,
    "LiI":    0.1797,
    "LiOH":   0.0894,
    "LiNO3":  0.1820,
    "Li2SO4": 0.2936,
    "LiClO3": 0.1325,
    "CaCl2":  0.2170,
}
_A4_DEFAULT = 0.1250


class MediumRange:
    """Computes ln(γ_MR) for all four species."""

    # ── public API ────────────────────────────────────────────────────────────

    def func_MR(
        self,
        b_ij: np.ndarray,        # shape (4,)
        c_ij: np.ndarray,        # shape (4,)
        b_jcja: float,
        c_jcja: float,
        x: np.ndarray,           # shape (N,4)
        MW: np.ndarray,          # shape (4,)
        valency: np.ndarray,     # shape (4,)
        salt: str,
    ) -> np.ndarray:             # shape (N,4)
        """Medium-range ln(γ), no explicit temperature dependence."""
        m_ion, ionic_strength, x_saltfree, mean_MW = self._variables(x, MW, valency)
        B_solv, dB_solv, B_ion, dB_ion = self._interaction_terms(
            b_ij, c_ij, b_jcja, c_jcja, ionic_strength, salt
        )
        ln_gamma_solv = self._solvent_gamma(
            m_ion, x_saltfree, ionic_strength, MW, mean_MW,
            B_solv, dB_solv, B_ion, dB_ion
        )
        ln_gamma_ions = self._ions_gamma(
            mean_MW, m_ion, x_saltfree,
            B_solv, dB_solv, B_ion, dB_ion, MW, b_ij, c_ij, valency
        )
        return np.hstack([ln_gamma_solv, ln_gamma_ions])

    def func_MR_T(
        self,
        b_ij: np.ndarray,
        c_ij: np.ndarray,
        b_jcja: float,
        c_jcja: float,
        x: np.ndarray,
        MW: np.ndarray,
        valency: np.ndarray,
        salt: str,
        T: np.ndarray,           # shape (N,)
    ) -> np.ndarray:
        """Medium-range ln(γ) with temperature dependence in B terms."""
        m_ion, ionic_strength, x_saltfree, mean_MW = self._variables(x, MW, valency)
        B_solv, dB_solv, B_ion, dB_ion = self._interaction_terms_T(
            b_ij, c_ij, b_jcja, c_jcja, ionic_strength, T, salt
        )
        ln_gamma_solv = self._solvent_gamma(
            m_ion, x_saltfree, ionic_strength, MW, mean_MW,
            B_solv, dB_solv, B_ion, dB_ion
        )
        ln_gamma_ions = self._ions_gamma(
            mean_MW, m_ion, x_saltfree,
            B_solv, dB_solv, B_ion, dB_ion, MW, b_ij, c_ij, valency
        )
        return np.hstack([ln_gamma_solv, ln_gamma_ions])

    # ── private helpers ───────────────────────────────────────────────────────

    def _variables(self, x, MW, valency):
        kg_solvent   = x[:, 0] * MW[0] + x[:, 1] * MW[1]
        m_ion        = x[:, 2:4] / kg_solvent[:, np.newaxis]
        ionic_strength = 0.5 * np.sum(m_ion * valency[2:4]**2, axis=1)
        x_saltfree   = np.column_stack([
            x[:, 0] / (x[:, 0] + x[:, 1]),
            x[:, 1] / (x[:, 0] + x[:, 1]),
        ])
        mean_MW = x_saltfree[:, 0] * MW[0] + x_saltfree[:, 1] * MW[1]
        return m_ion, ionic_strength, x_saltfree, mean_MW

    @staticmethod
    def _a_constants(salt: str):
        a1 = -1.2
        a3 = -1.0
        a4 = _A4_MAP.get(salt, _A4_DEFAULT)
        a2 = 2 * a4
        return a1, a2, a3, a4

    def _interaction_terms(self, b_ij, c_ij, b_jcja, c_jcja, ionic_strength, salt):
        a1, a2, a3, a4 = self._a_constants(salt)
        sqrt_I = np.sqrt(ionic_strength)
        exp_solv = np.exp(a1 * sqrt_I + a2 * ionic_strength)
        exp_ion  = np.exp(a3 * sqrt_I + a4 * ionic_strength)

        # B_solv shape: (N,4)  [solv-cat, solv-an, water-cat, water-an]
        B_solv  = b_ij + c_ij * exp_solv[:, np.newaxis]
        dB_solv = (a1 / 2.0 / sqrt_I + a2)[:, np.newaxis] * c_ij * exp_solv[:, np.newaxis]

        # B_ion shape: (N,)
        B_ion  = b_jcja + c_jcja * exp_ion
        dB_ion = (a3 / 2.0 / sqrt_I + a4) * c_jcja * exp_ion
        return B_solv, dB_solv, B_ion, dB_ion

    def _interaction_terms_T(self, b_ij, c_ij, b_jcja, c_jcja, ionic_strength, T, salt):
        a1, a2, a3, a4 = self._a_constants(salt)
        sqrt_I = np.sqrt(ionic_strength)
        exp_solv = np.exp(a1 * sqrt_I + a2 * ionic_strength)
        exp_ion  = np.exp(a3 * sqrt_I + a4 * ionic_strength)

        B_solv  = b_ij / T[:, np.newaxis] + c_ij / T[:, np.newaxis] * exp_solv[:, np.newaxis]
        dB_solv = (a1 / 2.0 / sqrt_I + a2)[:, np.newaxis] * c_ij / T[:, np.newaxis] * exp_solv[:, np.newaxis]

        B_ion  = b_jcja / T + c_jcja / T * exp_ion
        dB_ion = (a3 / 2.0 / sqrt_I + a4) * c_jcja / T * exp_ion
        return B_solv, dB_solv, B_ion, dB_ion

    def _solvent_gamma(self, m_ion, x_saltfree, ionic_strength, MW, mean_MW,
                       B_solv, dB_solv, B_ion, dB_ion):
        # Shared intermediate: sum over ions of x_sf * m * (B + I*dB)
        shared = (
            np.sum(x_saltfree[:, 0:1] * m_ion * (B_solv[:, 0:2] + ionic_strength[:, np.newaxis] * dB_solv[:, 0:2]), axis=1)
          + np.sum(x_saltfree[:, 1:2] * m_ion * (B_solv[:, 2:4] + ionic_strength[:, np.newaxis] * dB_solv[:, 2:4]), axis=1)
        )

        Solv1_T1 = np.sum(m_ion * B_solv[:, 0:2], axis=1)
        Solv1_T2 = MW[0] / mean_MW * shared

        Solv2_T1 = np.sum(m_ion * B_solv[:, 2:4], axis=1)
        Solv2_T2 = MW[1] / mean_MW * shared

        Solv_T3  = m_ion[:, 0] * m_ion[:, 1] * (B_ion + ionic_strength * dB_ion)

        ln_g1 = Solv1_T1 - Solv1_T2 - MW[0] * Solv_T3
        ln_g2 = Solv2_T1 - Solv2_T2 - MW[1] * Solv_T3
        return np.column_stack([ln_g1, ln_g2])

    def _ions_gamma(self, mean_MW, m_ion, x_saltfree,
                    B_solv, dB_solv, B_ion, dB_ion,
                    MW, b_ij, c_ij, valency):
        # Term 1: weighted sum of B_solv over salt-free fractions
        Cat_T1 = (1.0 / mean_MW
                  * (x_saltfree[:, 0] * B_solv[:, 0] + x_saltfree[:, 1] * B_solv[:, 2]))
        An_T1  = (1.0 / mean_MW
                  * (x_saltfree[:, 0] * B_solv[:, 1] + x_saltfree[:, 1] * B_solv[:, 3]))

        # Term 2: valency²/(2*mean_MW) × sum of x_sf * m * dB_solv
        # v_sq shape: (2,), mean_MW shape: (N,), shared_dB shape: (N,)
        v_sq = valency[2:4]**2
        shared_dB = (
            np.sum(x_saltfree[:, 0:1] * m_ion * dB_solv[:, 0:2], axis=1)
          + np.sum(x_saltfree[:, 1:2] * m_ion * dB_solv[:, 2:4], axis=1)
        )
        # Result shape: (N, 2)
        Ions_T2 = (v_sq[np.newaxis, :] / (2.0 * mean_MW[:, np.newaxis])) * shared_dB[:, np.newaxis]

        # Term 3: counter-ion molality × B_ion
        Ions_T3 = np.column_stack([
            m_ion[:, 1] * B_ion,
            m_ion[:, 0] * B_ion,
        ])

        # Term 4: 0.5 * z² * m_cat * m_an * dB_ion
        Ions_T4 = (0.5 * v_sq[np.newaxis, :]
                   * (m_ion[:, 0] * m_ion[:, 1] * dB_ion)[:, np.newaxis])

        # Reference states (infinite dilution in pure water)
        Ions_Ref_Cat = (1.0 / MW[1]) * (b_ij[2] + c_ij[2])
        Ions_Ref_An  = (1.0 / MW[1]) * (b_ij[3] + c_ij[3])

        ln_g_cat = Cat_T1 + Ions_T2[:, 0] + Ions_T3[:, 0] + Ions_T4[:, 0] - Ions_Ref_Cat
        ln_g_an  = An_T1  + Ions_T2[:, 1] + Ions_T3[:, 1] + Ions_T4[:, 1] - Ions_Ref_An
        return np.column_stack([ln_g_cat, ln_g_an])